core/src/storage/collections/binary_heap/duplex_sync_chunk.rs

```diff
@@ -0,0 +1,194 @@
+// Copyright 2018-2019 Parity Technologies (UK) Ltd.
+// This file is part of ink!.
+//
+// ink! is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// ink! is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with ink!.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Provides a wrapper around `SyncChunk` which stores a defined amount
+//! of values in one cell (instead of the `SyncChunk` behavior of storing
+//! one value per cell). The intention is to reduce expensive fetch
+//! operations from storage.
+//!
+//! **NOTE** This wrapper is geared explicitly towards a binary tree
+//! structure -- the value which is stored at index `0` (the root) will
+//! always be stored in its own group with no other values in it. The
+//! intention is to be able to store child nodes paired together
+//! in a group, since for query operations you have to access both
+//! elements anyways. This allows to skip one expensive read for every
+//! accessed pair.
+//!
+//! For example, for `COUNT = 2` the first group (at index `0`) will
+//! contain `[Some(root), None]`. The subsequent group at index `1`
+//! will contain `[Some(value from index 1), Some(value from index 2)]`.
+//! The getters and setters exposed by this module take care of mapping
+//! to the correct group index.
+
+use crate::storage::{
+    chunk::SyncChunk,
+    Flush,
+};
+use scale::{
+    Decode,
+    Encode,
+};
+
+// Number of values stored in each entry of the `SyncChunk`.
+// Note that the first group (at index `0`) will only ever
+// contain one value.
+const COUNT: u32 = 2;
```

Comment:
To make the implementation more generic and more usable with respect to other components that might need other values for `COUNT` we could make `Group` generic over array-like instances `[(); N]` where `N` reflects `COUNT` and by offering a trait `Count { const N: usize; }` that is auto implemented for `[(); 0..=32]` we could go a long way I guess. But then again I am not sure if Rust's current const-eval capabilities are strong enough for such a use-case. We would need to research. Also I think this is better suited in a follow-up PR. :)

Reply:
That's a really good point, I was already wondering how we could make it generic, while still having the `COUNT` determined at compile time. I'll try to take that into account when reusing this module for `BTreeMap`.